### DISCUSSION TITLE
Question about how `java.net.http.HttpClient#sendAsync` instrumentation works

### DISCUSSION THREAD

**Discussion by fmterrorf** [open-telemetry/opentelemetry-java-instrumentation#6780] [Q&A] [answered]
Hello :wave: I'm very curious on how otel is able to hook into `sendAsync` https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d919f84cf80738e7b6c375dcca3edd5cb531fe49/instrumentation/java-http-client/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/httpclient/HttpClientInstrumentation.java#L58. 

I'm using bytebuddy to create an agent but for some reason the binconsumer that is supplied in `future.whenComplete`https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/d919f84cf80738e7b6c375dcca3edd5cb531fe49/instrumentation/java-http-client/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/httpclient/HttpClientInstrumentation.java#L136 throws a `java.lang.NoClassDefFoundError`.

**Comment by laurit** [comment, marked as answer]
You need to ensure that the class you instrument (in this case `jdk.internal.net.http.HttpClientImpl`) has access to the class that you use in your advice (lets call it `ResponseConsumer`). The first hurdle is that the javaagent is loaded with app class loader (system class loader in jdk8) and `HttpClientImpl` is loaded with platform class loader (it is not present in jdk8, but on jdk8 jdk classes are loaded by boot loader). Platform loader is parent of app loader so `HttpClientImpl` won't have access to classes in agent jar. The way you get around this is adding your agent to boot loader with [`Instrumentation.appendToBootstrapClassLoaderSearch`](https://docs.oracle.com/en/java/javase/14/docs/api/java.instrument/java/lang/instrument/Instrumentation.html#appendToBootstrapClassLoaderSearch(java.util.jar.JarFile)). Typically you'd do this as the verify first thing in the agent to ensure that all agent classes (except the premain class) are in boot loader. We do this in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/javaagent-bootstrap/src/main/java/io/opentelemetry/javaagent/OpenTelemetryAgent.java
The next problem that you'll probably hit is that your agent classes are in unnamed module of boot loader but `HttpClientImpl` doesn't read that module and can't use your class. You should be able to get around this with `--add-reads=java.net.http=ALL-UNNAMED` (didn't try this). Same can be done programmatically with [`Instrumentation.redefineModule`](https://docs.oracle.com/en/java/javase/14/docs/api/java.instrument/java/lang/instrument/Instrumentation.html#redefineModule(java.lang.Module,java.util.Set,java.util.Map,java.util.Map,java.util.Set,java.util.Map)) or with bytebuddy. We do this in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/javaagent-tooling/javaagent-tooling-java9/src/main/java/io/opentelemetry/javaagent/tooling/ExposeAgentBootstrapListener.java
In our agent the classes that are used in advice are packaged in a subdirectory inside the agent jar and they can't be directly loaded so we need to define them in the appropriate loader ourself. This is handled in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/muzzle/src/main/java/io/opentelemetry/javaagent/tooling/HelperInjector.java

**Comment by fmterrorf** [comment]
Thank you very much for the answer
